tchap_red_list/__init__.py

```diff
@@ -11,36 +11,211 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
-from typing import Any, Dict
+import logging
+from typing import Any, Dict, Optional, Tuple
 
 import attr
-from synapse.module_api import ModuleApi
+from synapse.module_api import (
+    DatabasePool,
+    JsonDict,
+    LoggingTransaction,
+    ModuleApi,
+    UserProfile,
+    cached,
+    run_in_background,
+)
+
+logger = logging.getLogger(__name__)
+
+ACCOUNT_DATA_TYPE = "im.vector.hide_profile"
 
 
 @attr.s(auto_attribs=True, frozen=True)
 class RedListManagerConfig:
-    pass
+    discovery_room: Optional[str] = None
 
 
 class RedListManager:
-    def __init__(self, config: RedListManagerConfig, api: ModuleApi):
+    def __init__(
+        self, config: RedListManagerConfig, api: ModuleApi, setup_db: bool = True
+    ):
         # Keep a reference to the config and Module API
         self._api = api
         self._config = config
 
+        # Register callbacks
+        self._api.register_account_data_callbacks(
+            on_account_data_updated=self.update_red_list_status,
+        )
+
+        self._api.register_spam_checker_callbacks(
+            check_username_for_spam=self.check_user_in_red_list,
+        )
+
+        if setup_db:
+            # Set up the storage layer
+            run_in_background(self._setup_db)
+
     @staticmethod
     def parse_config(config: Dict[str, Any]) -> RedListManagerConfig:
-        # Parse the module's configuration here.
-        # If there is an issue with the configuration, raise a
-        # synapse.module_api.errors.ConfigError.
-        #
-        # Example:
-        #
-        #     some_option = config.get("some_option")
-        #     if some_option is None:
-        #          raise ConfigError("Missing option 'some_option'")
-        #      if not isinstance(some_option, str):
-        #          raise ConfigError("Config option 'some_option' must be a string")
-        #
-        return RedListManagerConfig()
+        return RedListManagerConfig(**config)
+
+    async def update_red_list_status(
+        self,
+        user_id: str,
+        room_id: Optional[str],
+        account_data_type: str,
+        content: JsonDict,
+    ) -> None:
+        """Update a user's status in the red list when their account data changes.
+        Implements the on_account_data_updated account data callback.
+        """
+        if account_data_type != ACCOUNT_DATA_TYPE:
+            return
+
+        # Compare what status (in the list, not in the list) the user wants to have with
+        # what it already has. If they're the same, don't do anything more.
+        desired_status = content.get("hide_profile") or False
```

Comment:
```suggestion
        desired_status = bool(content.get("hide_profile"))
```
otherwise the `current_status == desired_status` condition below can have unexpected results when the `hide_profile` field isn't a boolean.